### PR TITLE
Use ManualResetEventSlim instead of TaskCompletionSource

### PR DESCRIPTION
**Pull request by ajaybhargavb**
https://github.com/aspnet/Razor/issues/1991

Doing this fixes the flakiness in two tests.

Better to review using https://github.com/aspnet/Razor/pull/1994/files?w=1

### CODE REVIEW

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs [hunk 1]
```diff
@@ -116,82 +116,86 @@ public void ServerStartup_SuccessfullyAcquiredMutex()
...
             {
-                // We want this to keep running even after the shutdown is seen.
-                completionSource.Task.Wait();
-                return EmptyServerResponse;
-            });
+                var host = CreateCompilerHost(c => c.ExecuteFunc = (req, ct) =>
+                {
+                    // We want this to keep running even after the shutdown is seen.
+                    startedMre.Set();
+                    finishedMre.Wait();
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs]: This should have some reasonable timeout

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs [hunk 2]
```diff
@@ -116,82 +116,86 @@ public void ServerStartup_SuccessfullyAcquiredMutex()
...
-                    Assert.Equal(Process.GetCurrentProcess().Id, processId);
-                    Assert.False(compileTask.IsCompleted);
-                }
+                    var compileTask = ServerUtilities.Send(serverData.PipeName, EmptyServerRequest);
+                    startedMre.Wait();
 
-                // Now let the task complete.
-                completionSource.SetResult(true);
+                    // Act
+                    for (var i = 0; i < 10; i++)
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs]: Why is this in a loop?

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs [hunk 3]
```diff
@@ -116,82 +116,86 @@ public void ServerStartup_SuccessfullyAcquiredMutex()
...
+                    {
+                        // The compilation is now in progress, send the shutdown.
+                        var processId = await ServerUtilities.SendShutdown(serverData.PipeName);
+                        Assert.Equal(Process.GetCurrentProcess().Id, processId);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs]: Out of curiosity, what's the significance of this assertion?

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs (hunk 2 above)
Comment (ajaybhargavb) [reply test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs]: This is a test for testing multiple shutdown requests.

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs (hunk 3 above)
Comment (ajaybhargavb) [reply test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs]: I think this is to make sure we shutdown the same process that we started and not some unrelated server that was running. Fyi these tests were ported from https://github.com/dotnet/roslyn/blob/master/src/Compilers/Server/VBCSCompilerTests/CompilerServerApiTest.cs#L387-L513

Comment (pranavkm) [reply test/Microsoft.AspNetCore.Razor.Tools.Test/ServerLifecycleTest.cs]: Could you add a comment to this effect? In the event this ever fails, it'll be more obvious what it is that failed.

**Comment by ajaybhargavb** [comment]
Closing this as the correct fix is here https://github.com/aspnet/Razor/pull/2128

**Pull request closed without merging**